src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+    let elems: Vec<Result<u16, _>> = s.split(':').map(|s| u16::from_str_radix(s, 16)).collect();
+    if elems.len() != 3 || !elems.iter().all(Result::is_ok) {
+        return None;
+    };
+
+    let numbers: Vec<u16> = elems.into_iter().map(|n| n.unwrap()).collect();
+    Some((numbers[0], numbers[1], numbers[2]))
+}
+
+/// Object for accessing the HID device
+pub struct HidDevice {
+    blocking: Cell<bool>,
+    fd: OwnedFd,
+    info: RefCell<Option<DeviceInfo>>,
+}
+
+unsafe impl Send for HidDevice {}
+
+// API for the library to call us, or for internal uses
+impl HidDevice {
+    pub(crate) fn open(vid: u16, pid: u16, sn: Option<&str>) -> HidResult<Self> {
+        for device in HidApiBackend::get_hid_device_info_vector()?
+            .iter()
+            .filter(|device| device.vendor_id == vid && device.product_id == pid)
+        {
+            match (sn, &device.serial_number) {
+                (None, _) => return Self::open_path(&device.path),
+                (Some(sn), WcharString::String(serial_number)) if sn == serial_number => {
+                    return Self::open_path(&device.path)
+                }
+                _ => continue,
+            };
+        }
+
+        Err(HidError::HidApiError {
+            message: "device not found".into(),
+        })
+    }
+
+    pub(crate) fn open_path(device_path: &CStr) -> HidResult<HidDevice> {
+        // Paths on Linux can be anything but devnode paths are going to be ASCII
+        let path = device_path.to_str().expect("path must be utf-8");
+        let fd: OwnedFd = match OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_CLOEXEC | libc::O_NONBLOCK)
+            .open(path)
+        {
+            Ok(f) => f.into(),
+            Err(e) => {
+                return Err(HidError::HidApiError {
+                    message: format!("failed to open device with path {path}: {e}"),
+                });
+            }
+        };
+
+        let mut size = 0_i32;
+        if let Err(e) = unsafe { hidraw_ioc_grdescsize(fd.as_raw_fd(), &mut size) } {
+            return Err(HidError::HidApiError {
+                message: format!("ioctl(GRDESCSIZE) error for {path}, not a HIDRAW device?: {e}"),
+            });
+        }
+
+        Ok(Self {
+            blocking: Cell::new(true),
+            fd,
+            info: RefCell::new(None),
+        })
+    }
+
+    fn info(&self) -> HidResult<Ref<DeviceInfo>> {
+        if self.info.borrow().is_none() {
+            let info = self.get_device_info()?;
+            self.info.replace(Some(info));
+        }
+
+        let info = self.info.borrow();
+        Ok(Ref::map(info, |i: &Option<DeviceInfo>| i.as_ref().unwrap()))
+    }
+}
+
+impl HidDeviceBackendBase for HidDevice {
+    fn write(&self, data: &[u8]) -> HidResult<usize> {
+        if data.is_empty() {
+            return Err(HidError::InvalidZeroSizeData);
+        }
+
+        Ok(write(self.fd.as_raw_fd(), data)?)
```

Comment:
Good question, this is what the C hidapi does so it might be fine? This isn't a regular file write but this ends up enqueuing into whatever queue the kernel has so I don't know if it this could block.

Reply:
Based on https://docs.kernel.org/hid/hidraw.html where it only mentions blocking on `read()` it sounds like this is more akin to putting stuff in a queue than dong IO.

Reply:
Ok interesting. Although its always a bit weird when special devices in linux do not behave like e.g. the read() or write() manpages describe it.

But what happens if the queue is full? Maybe it will block then. Could be a theoretical possibility, that has no chance of practically happening. Lets keep it like it is, if we ever have weird issues in regards to writing, we can look again.

Reply:
Looking through a bit of kernel code, it looks like the `hidraw_write` function does a bunch of checking and you just end up doing the same as the send-report ioctl, and then it gets handed over to the appropriate driver to whatever callback it has decided. So it looks like what happens is whatever each device driver decides to do.